README
NAME
    DBIx::Class::InflateColumn::Serializer - Inflators to serialize data
    structures for DBIx::Class

SYNOPSIS
      package MySchema::Table;
      use base 'DBIx::Class';

      __PACKAGE__->load_components('InflateColumn::Serializer', 'Core');
      __PACKAGE__->add_columns(
        'data_column' => {
          'data_type' => 'VARCHAR',
          'size'      => 255,
          'serializer_class'   => 'JSON' 
        }
      );

    Then in your code...

      my $struct = { 'I' => { 'am' => 'a struct' };
      $obj->data_column($struct);
      $obj->update;

    And you can recover your data structure with:

      my $obj = ...->find(...);
      my $struct = $obj->data_column;

    The data structures you assign to "data_column" will be saved in the
    database in JSON format.

DESCRIPTION
    These modules help you store and access serialized data structures in
    the columns of your DB from your DBIx::Classes. They are inspired from
    the DBIx::Class::Manual::FAQ and the DBIC test suite, and provide a bit
    more protection than the inflators proposed in the FAQ. The intention is
    to provide a suite of well proven and reusable inflators and deflators
    to complement DBIx::Class.

    Added features for these inflators are: - throw an exception if the
    serialization doesn't fit in the field - throw an exception if the
    deserialization results in an error

    Right now there are three serializers: - Storable - JSON - YAML

USAGE
    1. Choose your serializer: JSON, YAML or Storable

    2. Add 'InflateColumn::Serializer' into the load_components of your
    table class

    3. add 'serializer_class' => SERIALIZER to the properties of the column
    that you want to (de/i)nflate with the SERIALIZER class.

NOTES
    As stated in the DBIC FAQ: "Be careful not to overuse this capability,
    however. If you find yourself depending more and more on some data
    within the inflated column, then it may be time to factor that data
    out."

AUTHOR
        Jose Luis Martinez
        CPAN ID: JLMARTIN
        CAPSiDE
        jlmartinez@capside.com
        http://www.pplusdomain.net

COPYRIGHT
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.

THANKS
    Matt S Trout for his valuable feedback

    Ask Bjorn Hansen

    Karen Etheridge

SEE ALSO
    DBIx::Class, DBIx::Class::Manual::FAQ



Changes
Revision history for DBIx-Class-InflateColumn-Serializer

0.07      2014-11-06 20:34:13+00:00 Europe/London
    - Switch JSON backend JSON::MaybeXS. Thanks to ether for the PR

0.06      2013-06-28 18:36:02 Europe/Madrid
    - Ticket #86304 fixed. Thanks to ribasushi for the bug report :)

0.05      2013-06-03 22:44:09 Europe/Madrid
    - Not using GitHub issues for this module
    - Version in modules now managed by Dist::Zilla plugin

0.04      2013-05-23 23:02:23 Europe/Madrid
    - Some POD errors fixed
    - Author tests moved to xt/
    - Migrate to Dist::Zilla

0.03 Mon Feb 7 18:48:00 2011
    - Migrate to Module::Install
    - Declare test_requires and optional packages (solves RT #64925)
    - use JSON instead of JSON::Any (RT #61551)

0.02 Sat Jun 27 13:32:00 2009
    - Moved from DBIx::Class::Inflator:: to DBIx::Class::InflateColumn
    - Moved to a driver model for the serializers. No need to specify
      the serializer in the load_components, just InflateColumn::Serializer
    - abandonded is_json => 1, is_storable => 1, is_yaml => 1 properties for
      serializer_class => 'JSON', serializer_class => 'YAML',
      serializer_class => 'Storable'

0.01 Mon Jun  8 22:39:35 2009
    - original version


t/lib/DBICTest.pm
package # hide from PAUSE 
   DBICTest;

use strict;
use warnings;

use DBICTest::Schema;

sub populate_schema {

}

sub has_custom_dsn {
	return $ENV{"DBICTEST_DSN"} ? 1:0;
}

sub _sqlite_dbfilename {
    return "t/var/DBIxClass.db";
}

sub _sqlite_dbname {
    my $self = shift;
    my %args = @_;
    return $self->_sqlite_dbfilename if $args{sqlite_use_file} or $ENV{"DBICTEST_SQLITE_USE_FILE"};
	return ":memory:";
}

sub _database {
    my $self = shift;
    my %args = @_;
    my $db_file = $self->_sqlite_dbname(%args);

    unlink($db_file) if -e $db_file;
    unlink($db_file . "-journal") if -e $db_file . "-journal";
    mkdir("t/var") unless -d "t/var";

    my $dsn = $ENV{"DBICTEST_DSN"} || "dbi:SQLite:${db_file}";
    my $dbuser = $ENV{"DBICTEST_DBUSER"} || '';
    my $dbpass = $ENV{"DBICTEST_DBPASS"} || '';

    my @connect_info = ($dsn, $dbuser, $dbpass, { AutoCommit => 1, sqlite_unicode => 1, %args });

    return @connect_info;
}

sub init_schema {
    my $self = shift;
    my %args = @_;

    my $schema;
    
    $args{'no_deploy'} = $ENV{'DBICTEST_NODEPLOY'} if (defined $ENV{'DBICTEST_NODEPLOY'});

    if ($args{compose_connection}) {
      $schema = DBICTest::Schema->compose_connection(
                  'DBICTest', $self->_database(%args)
                );
    } else {
      $schema = DBICTest::Schema->compose_namespace('DBICTest');
    }
    if( $args{storage_type}) {
    	$schema->storage_type($args{storage_type});
    }    
    if ( !$args{no_connect} ) {
      $schema = $schema->connect($self->_database(%args));
      $schema->storage->on_connect_do(['PRAGMA synchronous = OFF'])
       unless $self->has_custom_dsn;
    }
    if ( !$args{no_deploy} ) {
        __PACKAGE__->deploy_schema( $schema, $args{deploy_args} );
        __PACKAGE__->populate_schema( $schema )
         if( !$args{no_populate} );
    }
    return $schema;
}

sub deploy_schema {
    my $self = shift;
    my $schema = shift;
    my $args = shift || {};

    if ($ENV{"DBICTEST_SQLT_DEPLOY"}) { 
        $schema->deploy($args);    
    } else {
        open IN, "t/lib/sqlite.sql";
        my $sql;
        { local $/ = undef; $sql = <IN>; }
        close IN;
        for my $chunk ( split (/;\s*\n+/, $sql) ) {
          if ( $chunk =~ / ^ (?! --\s* ) \S /xm ) {  # there is some real sql in the chunk - a non-space at the start of the string which is not a comment
            $schema->storage->dbh->do($chunk) or print "Error on SQL: $chunk\n";
          }
        }
    }
    return;
}

1;


1;


t/lib/DBICTest/Schema/TestTable.pm
package # hide from PAUSE
    DBICTest::Schema::TestTable;

use base qw/DBIx::Class/;

__PACKAGE__->load_components (qw/Core/);

__PACKAGE__->table('testtable');
__PACKAGE__->add_columns(
  'testtable_id' => {
    data_type => 'integer',
  },
  'serial1' => {
    data_type => 'varchar',
    size => 100,
    is_nullable => 1,
  },
  'serial2' => {
    data_type => 'varchar'
    is_nullable => 1,
  }
);

__PACKAGE__->set_primary_key('testtable_id');

1;



t/lib/DBICTest/Schema/SerializeJSON.pm
package # hide from PAUSE
    DBICTest::Schema::SerializeJSON;

use base qw/DBIx::Class/;

__PACKAGE__->load_components (qw/InflateColumn::Serializer Core/);

__PACKAGE__->table('testtable');
__PACKAGE__->add_columns(
  'testtable_id' => {
    data_type => 'integer',
  },
  'serial1' => {
    data_type => 'varchar',
    size => 100,
    serializer_class => 'JSON',
    is_nullable => 1,
  },
  'serial2' => {
    data_type => 'varchar',
    serializer_class => 'JSON',
    is_nullable => 1,
  },
  'serial3' => {
    data_type => 'varchar',
    serializer_class => 'JSON',
    serializer_options => { allow_nonref => 1 },
    is_nullable => 1,
  }
);

__PACKAGE__->set_primary_key('testtable_id');

1;



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.009.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Inflators to serialize data structures for DBIx::Class",
  "AUTHOR" => "Jose Luis Martinez (jlmartinez\@capside.com)",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "DBIx-Class-InflateColumn-Serializer",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "DBIx::Class::InflateColumn::Serializer",
  "PREREQ_PM" => {
    "DBIx::Class" => 0
  },
  "TEST_REQUIRES" => {
    "DBD::SQLite" => 0,
    "JSON::MaybeXS" => "1.002005",
    "Test::Exception" => 0,
    "Test::More" => 0,
    "YAML" => 0,
    "namespace::clean" => 0
  },
  "VERSION" => "0.07",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBD::SQLite" => 0,
  "DBIx::Class" => 0,
  "JSON::MaybeXS" => "1.002005",
  "Test::Exception" => 0,
  "Test::More" => 0,
  "YAML" => 0,
  "namespace::clean" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/DBIx/Class/InflateColumn/Serializer.pm
package DBIx::Class::InflateColumn::Serializer;
{
  $DBIx::Class::InflateColumn::Serializer::VERSION = '0.07';
}

use strict;
use warnings;

sub register_column {
    my $self = shift;
    my ($column, $info, $args) = @_;
    $self->next::method(@_);

    return unless defined $info->{'serializer_class'};


    my $class = "DBIx::Class::InflateColumn::Serializer::$info->{'serializer_class'}";
    eval "require ${class};";
    $self->throw_exception("Failed to use serializer_class '${class}': $@") if $@;

    defined( my $freezer = eval{ $class->get_freezer($column, $info, $args) }) ||
      $self->throw_exception("Failed to create freezer with class '$class': $@");
    defined( my $unfreezer = eval{ $class->get_unfreezer($column, $info, $args) }) ||
      $self->throw_exception("Failed to create unfreezer with class '$class': $@");

    $self->inflate_column(
        $column => {
            inflate => $unfreezer,
            deflate => $freezer,
        }
    );
};

=head1 NAME

DBIx::Class::InflateColumn::Serializer - Inflators to serialize data structures for DBIx::Class

=head1 SYNOPSIS

  package MySchema::Table;
  use base 'DBIx::Class';

  __PACKAGE__->load_components('InflateColumn::Serializer', 'Core');
  __PACKAGE__->add_columns(
    'data_column' => {
      'data_type' => 'VARCHAR',
      'size'      => 255,
      'serializer_class'   => 'JSON' 
    }
  );

Then in your code...

  my $struct = { 'I' => { 'am' => 'a struct' };
  $obj->data_column($struct);
  $obj->update;

And you can recover your data structure with:

  my $obj = ...->find(...);
  my $struct = $obj->data_column;

The data structures you assign to "data_column" will be saved in the database in JSON format.

=head1 DESCRIPTION

These modules help you store and access serialized data structures in the columns of your DB from your DBIx::Classes. They are inspired from the DBIx::Class::Manual::FAQ and the DBIC test suite, and provide a bit more protection than the inflators proposed in the FAQ. The intention is to provide a suite of well proven and reusable inflators and deflators to complement DBIx::Class.

Added features for these inflators are:
 - throw an exception if the serialization doesn't fit in the field
 - throw an exception if the deserialization results in an error

Right now there are three serializers:
 - Storable
 - JSON
 - YAML

=head1 USAGE

1. Choose your serializer: JSON, YAML or Storable

2. Add 'InflateColumn::Serializer' into the load_components of your table class

3. add 'serializer_class' => SERIALIZER to the properties of the column that you want to (de/i)nflate
   with the SERIALIZER class.

=head1 NOTES

As stated in the DBIC FAQ: "Be careful not to overuse this capability, however. If you find yourself depending more and more on some data within the inflated column, then it may be time to factor that data out."

=head1 AUTHOR

    Jose Luis Martinez
    CPAN ID: JLMARTIN
    CAPSiDE
    jlmartinez@capside.com
    http://www.pplusdomain.net

=head1 COPYRIGHT

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=head1 THANKS

Matt S Trout for his valuable feedback

Ask Bjorn Hansen

Karen Etheridge

=head1 SEE ALSO

DBIx::Class, DBIx::Class::Manual::FAQ

=cut

#################### main pod documentation end ###################

1;



lib/DBIx/Class/InflateColumn/Serializer/YAML.pm
package DBIx::Class::InflateColumn::Serializer::YAML;
{
  $DBIx::Class::InflateColumn::Serializer::YAML::VERSION = '0.07';
}

=head1 NAME

DBIx::Class::InflateColumn::Serializer::YAML - YAML Inflator

=head1 SYNOPSIS

  package MySchema::Table;
  use base 'DBIx::Class';

  __PACKAGE__->load_components('InflateColumn::Serializer', 'Core');
  __PACKAGE__->add_columns(
    'data_column' => {
      'data_type' => 'VARCHAR',
      'size'      => 255,
      'serializer_class'   => 'YAML'
    }
  );

Then in your code...

  my $struct = { 'I' => { 'am' => 'a struct' };
  $obj->data_column($struct);
  $obj->update;

And you can recover your data structure with:

  my $obj = ...->find(...);
  my $struct = $obj->data_column;

The data structures you assign to "data_column" will be saved in the database in YAML format.

=cut

use strict;
use warnings;
use YAML;
use Carp;

=over 4

=item get_freezer

Called by DBIx::Class::InflateColumn::Serializer to get the routine that serializes
the data passed to it. Returns a coderef.

=cut

sub get_freezer{
  my ($class, $column, $info, $args) = @_;

  if (defined $info->{'size'}){
      my $size = $info->{'size'};
      return sub {
        my $s = YAML::Dump(shift);
        croak "serialization too big" if (length($s) > $size);
        return $s;
      };
  } else {
      return sub {
        return YAML::Dump(shift); 
      };
  }
}

=item get_unfreezer

Called by DBIx::Class::InflateColumn::Serializer to get the routine that deserializes
the data stored in the column. Returns a coderef.

=back

=cut

sub get_unfreezer {
  return sub {
    return YAML::Load(shift);
  };
}


1;


lib/DBIx/Class/InflateColumn/Serializer/Storable.pm
package DBIx::Class::InflateColumn::Serializer::Storable;
{
  $DBIx::Class::InflateColumn::Serializer::Storable::VERSION = '0.07';
}

=head1 NAME

DBIx::Class::InflateColumn::Serializer::Storable - Storable Inflator

=head1 SYNOPSIS

  package MySchema::Table;
  use base 'DBIx::Class';

  __PACKAGE__->load_components('InflateColumn::Serializer', 'Core');
  __PACKAGE__->add_columns(
    'data_column' => {
      'data_type' => 'VARCHAR',
      'size'      => 255,
      'serializer_class'   => 'Storable'
    }
  );

Then in your code...

  my $struct = { 'I' => { 'am' => 'a struct' };
  $obj->data_column($struct);
  $obj->update;

And you can recover your data structure with:

  my $obj = ...->find(...);
  my $struct = $obj->data_column;

The data structures you assign to "data_column" will be saved in the database in Storable format.

=cut

use strict;
use warnings;
use Storable qw//;
use Carp;

=over 4

=item get_freezer

Called by DBIx::Class::InflateColumn::Serializer to get the routine that serializes 
the data passed to it. Returns a coderef.

=cut

sub get_freezer{
  my ($class, $column, $info, $args) = @_;

  if (defined $info->{'size'}){
      my $size = $info->{'size'};
      return sub {
        my $s = Storable::nfreeze(shift);
        croak "serialization too big" if (length($s) > $size);
        return $s;
      };
  } else {
      return sub {
        return Storable::nfreeze(shift);
      };
  }
}

=item get_unfreezer

Called by DBIx::Class::InflateColumn::Serializer to get the routine that deserializes
the data stored in the column. Returns a coderef.

=back

=cut


sub get_unfreezer {
  return sub {
    my $value = shift;
    # Storable returns undef if the datastructure couldn't be thawed.
    # Other deserializers throw exceptions, so we'll do the same.
    # If the column had a NULL value, then we return it (don't want to die)
    return undef if (not defined $value);
    my $s = Storable::thaw($value);
    croak "Storable couldn't thaw the value" if (not defined $s);
    return $s;
  };
}


1;


lib/DBIx/Class/InflateColumn/Serializer/JSON.pm
package DBIx::Class::InflateColumn::Serializer::JSON;
{
  $DBIx::Class::InflateColumn::Serializer::JSON::VERSION = '0.07';
}

=head1 NAME

DBIx::Class::InflateColumn::Serializer::JSON - JSON Inflator

=head1 SYNOPSIS

  package MySchema::Table;
  use base 'DBIx::Class';

  __PACKAGE__->load_components('InflateColumn::Serializer', 'Core');
  __PACKAGE__->add_columns(
    'data_column' => {
      'data_type' => 'VARCHAR',
      'size'      => 255,
      'serializer_class'   => 'JSON',
      'serializer_options' => { allow_blessed => 1, convert_blessed => 1, pretty => 1 },    # optional
    }
  );

Then in your code...

  my $struct = { 'I' => { 'am' => 'a struct' };
  $obj->data_column($struct);
  $obj->update;

And you can recover your data structure with:

  my $obj = ...->find(...);
  my $struct = $obj->data_column;

The data structures you assign to "data_column" will be saved in the database in JSON format.

Any arguments included in C<serializer_options> will be passed to the L<JSON::MaybeXS> constructor,
to be used by the JSON backend for both serializing and deserializing.

=cut

use strict;
use warnings;
use JSON::MaybeXS;
use Carp;
use namespace::clean;

=over 4

=item get_freezer

Called by DBIx::Class::InflateColumn::Serializer to get the routine that serializes
the data passed to it. Returns a coderef.

=cut

sub get_freezer{
  my ($class, $column, $info, $args) = @_;

  my $opts = $info->{serializer_options};

  my $serializer = JSON::MaybeXS->new($opts && %$opts ? %$opts: ());

  if (defined $info->{'size'}){
      my $size = $info->{'size'};

      return sub {
        my $s = $serializer->encode(shift);
        croak "serialization too big" if (length($s) > $size);
        return $s;
      };
  } else {
      return sub {
        return $serializer->encode(shift);
      };
  }
}

=item get_unfreezer

Called by DBIx::Class::InflateColumn::Serializer to get the routine that deserializes
the data stored in the column. Returns a coderef.

=back

=cut

sub get_unfreezer {
  my ($class, $column, $info, $args) = @_;

  my $opts = $info->{serializer_options};
  return sub {
    JSON::MaybeXS->new($opts && %$opts ? %$opts : ())->decode(shift);
  };
}


1;
